std/experimental/bigo.d

```diff
@@ -0,0 +1,399 @@
+module std.experimental.bigo;
+
+struct BigO
+{
+    import std.typecons : Tuple;
+    import std.algorithm : cartesianProduct, map, sum;
+    import std.array : array, empty, front, popFront;
+    import std.range : chain;
+
+    alias Atom = Tuple!(int, "id", double, "exp", double, "logExp");
+    private Atom[][] terms;
+
+    string toString() const @safe
+    {
+        if (terms.empty) return "O(1)";
+        string result = "O(";
```

Comment:
I would use `std.array.appender` here.

Plus, since you know the length of `terms` and it's sub arrays, you could create a pretty good guess for the size needed, and call `appender.reserve` at the start of the function.